### DISCUSSION TITLE
BUG: If chrome.js is missing throws "Could not execute "run" due to missing configuration."

### DISCUSSION THREAD

**Discussion by darkartswizard** [webdriverio/webdriverio#8590] [Technical Discussion]
To reproduce:
Rename (or delete) chrome.js in \config folder.
from terminal: npm run chrome

WDIO generate this error : "Could not execute "run" due to missing configuration. Would you like to create one?"
Attempts to create (overwrite) existing wdio.config.js file.
Does not resolve issue. 
Sends user in wrong direction to resolve.

All Google references to this error have unrelated solutions. 

from terminal: npm run bogus
WDIO generates correct message: npm ERR! Missing script: "bogus"

**Comment by christian-bromann** [comment]
> Rename (or delete) chrome.js in \config folder.

Which `chrome.js`?

> WDIO generates correct message: npm ERR! Missing script: "bogus"

This is not WDIO it's NPM that tells you that you run a script that doesn't exist. Not related to wdio.

**Reply by darkartswizard** [reply]
Correct.
 
The first example has a reference to "chrome.js" which is referenced in the package.json, but did not exist in the \config folder.
In this case WDIO reports it could not find the wdio.config.js file, which did exist, and tries to rebuild it.
It should instead report the chrome.js file is missing from the \config folder, instead of attempting to recreate the existing wdio.config.js file which was unrelated to the issue.

The second example is the good wording of an error that would help the user identify the root cause..
Since "bogus" is not mentioned in the package.json the correct message is displayed by node.

**Reply by christian-bromann** [reply]
> The second example is the good wording of an error that would help the user identify the root cause.

You would need to report that to the NPM cli though.

> instead of attempting to recreate the existing wdio.config.js file which was unrelated to the issue.

I understand what you are suggesting here. From the DX perspective we thought asking to create a config would be a neat thing. We definitely could improve the wording a little bit here. What would you suggest? Like include the config file into the error message?

**Comment by darkartswizard** [comment]
In my package I have a "scripts" section.
One references a chrome script that was missing:
"scripts": {
    "chrome": "wdio run config/chrome.js --suite smoke",
    
    
This is the chrome.js config file. Since it was missing it's detail would be moot.

const { config } = require('../wdio.conf');
const device = require('./mobile-devices');

exports.config = {
    ...config,
    ...{
        services: ['chromedriver'],
        screenshotPath: `./errorShots`,
        capabilities: [
            {
                maxInstances: 5,
                browserName: 'chrome',
                'goog:chromeOptions': {
                    mobileEmulation: {
                        deviceName: device[process.env.DEVICE]
                    },
                    args: [
                        //'--headless',
                        '--window-size=1200,700'
                    ]
                }
            }
        ]
    }
}

**Comment by darkartswizard** [comment]
Sure. First I have no argument that creating the wdio.config.js is a neat
thing. It works really well.

But that is only cool if the wdio.config.js is Actually Missing. In this
case, the wdio.config.js exists, but WDIO insists it does not.

The problem in this edge case is that the cool thing is being triggered
when the wdio.config does exists, updates the existing wdio.config.js file,
and then upon re-running, states it still does not exist.

That’s the totally not cool thing. It is incorrectly reporting the wrong
config file is missing.

What I would suggest: Check for wdio.config.js existence, and if it
**does** exist, do not say “The wdio.config.js file is missing. Do you want
to create one?”

Instead check for the existence of the associated config file under
scripts:[…]( in this case .\config\chrome.js) and if that does not exist,
pop a message saying “the .\conf\config.js file is missing”.

I really don’t need to contact Node in the second example because it has no
issue.

I was trying to see what message shows when no path reference is in the
Package file to the .\config path. Chrome has a reference and Bogus does
not.

Node correctly show a message: “Missing Script: Bogus.”

**Comment by darkartswizard** [comment]
<img width="1035" alt="image" src="https://user-images.githubusercontent.com/50789857/180896138-d8edd743-cd46-4b16-a6f1-1b1770d45627.png">

Here is what happens when the "Bogus" config file exists and is referenced in the package file.
Work perfectly.

<img width="1059" alt="image" src="https://user-images.githubusercontent.com/50789857/180896517-6228b1fa-ce53-4065-8d32-7176e35470c2.png">

In this example I have renamed the bogus file. It is still referenced in the package.json.
But WDIO says the config file (wdio.config.js) is missing and asks to build the file.

It should be stating the \config\bogus.js file is missing, not the wdio.config.js file.

Does that make the issue clearer?

**Reply by darkartswizard** [reply]
<img width="984" alt="image" src="https://user-images.githubusercontent.com/50789857/180897991-9b3f185d-7ea4-4a1a-8954-cece2d2436e3.png">

Of course if I remove the reference, node give the correct error message with a good clue to where the issue resides.
Missing script: "Bogus"

They could say "Missing script: "Bogus" in the package.json file" but from that standpoint it's redundant.

**Reply by christian-bromann** [reply]
> It should be stating the \config\bogus.js file is missing, not the wdio.config.js file.

WebdriverIO also doesn't say that the `wdio.conf.js` is missing, it says __due to missing configuration__ which is correct as the configuration file you passed in is indeed missing. I get your point though that WebdriverIO could be more specific.

**Comment by christian-bromann** [comment]
Created https://github.com/webdriverio/webdriverio/issues/8592 .. any contributions are appreciated, thanks!

**Comment by darkartswizard** [comment]
I don’t know what to say.

Definitely meets the definition of Bad Code Smell when the tool says the
wrong file does not exist, overwrites the file, and then says that the file
still does not exist.

Could not agree that a message that fails to document exactly which missing
config file is at issue is good enough.

I can’t say it would be an easy fix, but certainly a something to consider
when removing obstacles to acceptance with new users.

It’s a pretty easy mistake to have a misspelled file name, or break a
relative path along the way.

But I gave a shout out to my dev team member who figured it out after
chasing my tail most of my Friday and Saturday.

I put it in my documentation should anyone here run across it again.

**Reply by christian-bromann** [reply]
> Definitely meets the definition of Bad Code Smell when the tool says the wrong file does not exist, overwrites the file, and then says that the file still does not exist.

Well, I disagree with your observation. WebdriverIO doesn't say a certain file does not exist. It says a configuration file can not be found based on provided parameters. Therefore it is a user error but the developer experience can be improved by clearly stating which file does not exist.

Let me clarify: the CLI error message says: "Could not execute 'run' due to missing configuration". It does __not__ mention the `wdio.conf.js` in the root directory at all. You passed in a path to a config file that doesn't exist and the tool failed due to that missing file.

> Could not agree that a message that fails to document exactly which missing config file is at issue is good enough.

Do you have a better suggestion?

**Reply by erwinheitzman** [reply]
Let me start by saying that this is a good catch.

I support what Christian has said and want to add some to that.

Perhaps we can additionally create the config using the path provided.
So if one runs `wdio run config/bogus.js` and we say something like `Could not execute 'run' due to missing configuration file config/bogus.js`, we could use the path to generate a new config at that location (`config/bogus.js`) by asking something like `Would you like to generate a config file at config/bogus.js?`.

When the file already exists we can ask if the user is sure it wants to overwrite the file.

This could be done in two stages:
1. change the message to contain the path
2. make the implementation a bit smarter

Regarding accidentally overwriting a config file, at this point it is a user error. Shouldn't be an issue when using source control.

What do you think?

**Comment by darkartswizard** [comment]
I would advocate for a simple message change today over a complex solution
in the future.

In my experience, I would follow the principle similar to Least Access
Privilege in Network Security.

The least we can do is tell the user exactly which config file is missing.
Just make the message detail smarter.

That would have saved me hours of time spinning my wheels. And I’m sure
I’ve been a bewildered victim to this before when I was first started on
WDIO projects last year.

Making this smarter by skipping the erroneous update to the wdio.conf.js
file would be great.

But it can be secondary to providing a quick fix and could be implemented
at a later date when all the details have been hammered out.
